packages/payment-detection/src/erc777/superfluid-retriever.ts

```diff
@@ -0,0 +1,101 @@
+  private client: TheGraphSuperfluidClient;
+
+  /**
+   * @param paymentReference The reference to identify the payment
+   * @param tokenContractAddress The address of the ERC777 contract
+   * @param toAddress Address of the balance we want to check
+   * @param eventName Indicate if it is an address for payment or refund
+   * @param network The Ethereum network to use
+   * @param options Extra options to GraphQL client
+   */
+  constructor(
+    private paymentReference: string,
+    private tokenContractAddress: string | null,
+    private toAddress: string,
+    private eventName: PaymentTypes.EVENTS_NAMES,
+    private network: string,
+    private options?: TheGraphClientOptions,
+  ) {
+    this.client = getTheGraphSuperfluidClient(this.network, this.options);
+  }
+
+  private getGraphVariables(): GraphPaymentQueryParams {
+    return {
+      reference: ethersUtils.keccak256(`0x${this.paymentReference}`),
+      to: this.toAddress,
+      tokenAddress: this.tokenContractAddress,
+    };
+  }
+
+  /**
+   * First MVP version which convert :
+   * stream events queried from SuperFluid subgraph
+   * into payment events with the parameters expected by extractEvents function
+   * to compute balance from amounts in ERC20 style transactions
+   */
+  public async getTransferEvents(): Promise<PaymentTypes.ERC20PaymentNetworkEvent[]> {
+    const variables = this.getGraphVariables();
+    const { flow, untagged } = await this.client.GetSuperFluidEvents(variables);
+    // Chronological sorting of events having payment reference and closing events without payment reference
+    const streamEvents = flow.concat(untagged).sort((a, b) => a.timestamp - b.timestamp);
+    const paymentEvents: PaymentTypes.ERC20PaymentNetworkEvent[] = [];
+    if (streamEvents.length < 1) {
+      return paymentEvents;
+    }
+
+    // if last event is ongoing stream then create end of stream to help compute balance
+    if (streamEvents[streamEvents.length - 1].flowRate > 0) {
+      streamEvents.push({
+        oldFlowRate: streamEvents[streamEvents.length - 1].flowRate,
+        flowRate: 0,
+        timestamp: Utils.getCurrentTimestampInSecond(),
+        blockNumber: null,
+        transactionHash: null,
+      } as FlowUpdatedEvent);
+    }
+
+    for (let index = 1; index < streamEvents.length; index++) {
+      // we have to manage update of flowrate to pay different payment references with the same token
+      // so we should care only about pairs of begin or update event (type 0 or 1) followed by end or update event (type 2 or 1)
+      // for each update of static flowrate between these 2 chronological sorted events:
+      // amount paid is the difference of previous flowrate multiplied by the difference of time
+      if (streamEvents[index - 1].type === 2 || streamEvents[index].type === 0) {
+        continue;
+      }
+      const amount =
+        (streamEvents[index].oldFlowRate - streamEvents[index - 1].oldFlowRate) *
```

Comment:
```suggestion
        (streamEvents[index].oldFlowRate - streamEvents[index - 1].flowRate) *
```
If I'm not wrong, it should mean that your tests are false positive.

Reply:
I've fixed it my local commit, it works for most tests because streamEvents[index].oldFlowRate is equal to streamEvents[index - 1].flowRate) in all my tests. So I should add one more when it's not in this case.